vault/data_source_ldap_dynamic_role_credentials.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package vault
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-provider-vault/internal/consts"
+	"github.com/hashicorp/terraform-provider-vault/internal/provider"
+	"github.com/hashicorp/vault/api"
+)
+
+func ldapDynamicCredDataSource() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: ReadContextWrapper(readLDAPDynamicCreds),
+		Schema: map[string]*schema.Schema{
+			consts.FieldMount: {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "LDAP Secret Backend to read credentials from.",
+			},
+			consts.FieldRoleName: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Name of the role.",
+			},
+			consts.FieldLeaseID: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Lease identifier assigned by Vault.",
+			},
+			consts.FieldLeaseDuration: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Lease duration in seconds.",
+			},
+			consts.FieldLeaseRenewable: {
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "True if the duration of this lease can be extended through renewal.",
+			},
+			consts.FieldDistinguishedNames: {
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "List of the distinguished names (DN) created.",
+				Elem:        &schema.Schema{Type: schema.TypeString},
+			},
+			consts.FieldPassword: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Password for the dynamic role.",
+				Sensitive:   true,
+			},
+			consts.FieldUsername: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Name of the dynamic role.",
+			},
+		},
+	}
+}
+
+func readLDAPDynamicCreds(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, err := provider.GetClient(d, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	mount := d.Get(consts.FieldMount).(string)
+	role := d.Get(consts.FieldRoleName).(string)
+	fullPath := fmt.Sprintf("%s/creds/%s", mount, role)
+
+	secret, err := client.Logical().ReadWithContext(ctx, fullPath)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error reading from Vault: %s", err))
+	}
+	log.Printf("[DEBUG] Read %q from Vault", fullPath)
+	if secret == nil {
+		return diag.FromErr(fmt.Errorf("no role found at %q", fullPath))
+	}
+
+	response, err := parseLDAPDynamicCredSecret(secret)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(response.username)
```

Comment:
We should set the ID to the `response.leaseID` here since we have that unique and important lease info readily available here. Since we didn't have that info available for the static roles, I wanted to check in on how we are setting the ID for that data source on [this line ](https://github.com/hashicorp/terraform-provider-vault/blob/VAULT-8099/secrets/ldap/vault/data_source_ldap_static_role_credentials.go#L99)

@raymonstah @fairclothjm do we want to set the ID to a more unique combination of values? I realized that it could be possible that two IDs would clash if the credentials are read for the same username (or is that not a concern?). Wanted to get your thoughts on if there was potentially a more unique choice for the ID in case of static role credentials

Reply:
ah yeah, using the `leaseID` seems preferable; i can update that.

as for the static credentials, I'm not sure if there is something unique we could use. I'm also not sure if this is a concern. could you explain the significance of setting the ID?

Reply:
The IDs are how Terraform keeps track of all the terraform resource/datasources. These IDs should be unique, and you can [see examples](https://github.com/hashicorp/terraform-provider-vault/blob/VAULT-8099/secrets/ldap/vault/resource_ldap_secret_backend.go#L209-L212) of how removing these IDs from the TF state can cause the terraform resource to be destroyed and cleaned up.

Causing two resources/data sources to have the same ID can cause either to be overwrited or a mismanagement of the TF state for that object, so we should avoid any ID clashes for resources and data sources. Let me know if you have any other questions!